textx/metamodel.py

```diff
@@ -547,32 +548,52 @@ def _current_namespace(self):
         return self.namespaces[self._namespace_stack[-1]]
 
     def model_from_str(self, model_str, file_name=None, debug=None,
-                       pre_ref_resolution_callback=None, encoding='utf-8'):
+                       pre_ref_resolution_callback=None, encoding='utf-8',
+                       **kwargs):
         """
         Instantiates model from the given string.
         :param pre_ref_resolution_callback: called before references are
                resolved. This can be useful to manage models distributed
                across files (scoping)
+        :param **kwargs additional arguments available through
+                _tx_model_kwargs after initiating the model. The attribute
+                is set while executing pre_ref_resolution_callback (see
+                scoping.md)
         """
 
+        def kwargs_callback(the_model):
+            if hasattr(the_model, '_tx_metamodel'):
+                the_model._tx_model_kwargs = ModelKwargs(kwargs)
+            if pre_ref_resolution_callback:
+                pre_ref_resolution_callback(the_model)
+
         if type(model_str) is not text:
             raise TextXError("textX accepts only unicode strings.")
 
         if file_name is None:
             model = self._parser_blueprint.clone().get_model_from_str(
                 model_str, debug=debug,
-                pre_ref_resolution_callback=pre_ref_resolution_callback)
+                pre_ref_resolution_callback=kwargs_callback)
 
             for p in self._model_processors:
                 p(model, self)
         else:
-            model = self.internal_model_from_file(file_name, encoding, debug,
-                                                  model_str=model_str)
+            model = self.internal_model_from_file(
+                file_name, encoding, debug,
+                model_str=model_str,
+                pre_ref_resolution_callback=kwargs_callback)
```

Comment:
Hmm. This is new: before this change, no callback was passed in that case. Unclear to me (but all tests pass).

Reply:
Fixed: reverted change and preserved 100% backwards compatibility.